### PR TITLE
MDEV-9911: NTILE must return an error when parameter is not stable

### PR DESCRIPTION
**Pull request by DanAtSPB**
We save NTILE argument(N) and compare it to the following N.
If N not a constant for the partition - we stop with an error.
Currently used an old error:Argument of NTILE must be greater than 0

### CODE REVIEW

File: sql/item_windowfunc.h [hunk 1]
```diff
@@ -635,11 +636,12 @@ class Item_sum_ntile : public Item_sum_window_with_row_count
 
     longlong num_quantiles= get_num_quantiles();
 
-    if (num_quantiles <= 0) {
+    if (num_quantiles <= 0 || (num_quantiles != n_old_val_ && n_old_val_ != 0))
```
Comment (an3l) [inline sql/item_windowfunc.h:639]: Please change condition `n_old_val_ != 0` to `n_old_val > 0` or please elaborate when can be `n_old_val_ <0` ?

File: mysql-test/t/win_ntile.test [hunk 2]
```diff
@@ -168,4 +168,55 @@ select pk, a, b,
...
+ (20, 'Lucky', 'Male', 5);
+--echo # Correct usage of NTILE with a fix argument NTILE(6). 
+select c1, c2, c3, ntile(6) over (partition by c2 order by c1) from t1;
+--echo # Correct usage - constant NTILE (argument) in each partition. 
+select c1, c2, c3, ntile(c3) over (partition by c2 order by c1) from t1;
+select c1, c2, c3, ntile(c3) over (partition by c1 order by c2) from t1;
+
+update t1 set c3= 1 where pk <= 10;
+
+--echo # We save NTILE argument(N) and compare it to the following N.
```
Comment (an3l) [inline mysql-test/t/win_ntile.test:210]: Three `echo` lines here are not needed.
If you want to explain better to do so in the code.

File: mysql-test/t/win_ntile.test [hunk 3]
```diff
@@ -168,4 +168,51 @@ select pk, a, b,
 from t1;
 
 
+drop table t1;
+
+--echo #
+--echo # MDEV-9911 NTILE must return an error when parameter is not stable
```
Comment (an3l) [inline mysql-test/t/win_ntile.test:174]: Here is whitespace at the end, no need.
Please take a look of other white space cases.
You can set your editor to show you whitespace changes when writing a patch.

File: mysql-test/t/win_ntile.test [hunk 4]
```diff
@@ -168,4 +168,51 @@ select pk, a, b,
...
+--echo #
+
+create table t1 (
+  pk int primary key,
+  c1 nvarchar(10),
+  c2 nvarchar(10),
+  c3 int
+);
+
+insert into t1 values
```
Comment (an3l) [inline mysql-test/t/win_ntile.test:184]: Whitespace

File: mysql-test/t/win_ntile.test [hunk 5]
```diff
@@ -168,4 +168,51 @@ select pk, a, b,
...
+ (12, 'John', 'Male', 5),
+ (13, 'Pam', 'Female', 6),
+ (14, 'Sara', 'Female', 6),
+ (15, 'Todd', 'Male', 5),
+ (16, 'Mary', 'Female', 6),
+ (17, 'Ben', 'Male', 5),
+ (18, 'Jodi', 'Female', 6),
+ (19, 'Tom', 'Male', 5),
+ (20, 'Lucky', 'Male', 5);
+--echo # Correct usage of NTILE with a fix argument NTILE(6).
```
Comment (an3l) [inline mysql-test/t/win_ntile.test:205]: Whitespace

File: mysql-test/t/win_ntile.test [hunk 6]
```diff
@@ -168,4 +168,51 @@ select pk, a, b,
...
+ (14, 'Sara', 'Female', 6),
+ (15, 'Todd', 'Male', 5),
+ (16, 'Mary', 'Female', 6),
+ (17, 'Ben', 'Male', 5),
+ (18, 'Jodi', 'Female', 6),
+ (19, 'Tom', 'Male', 5),
+ (20, 'Lucky', 'Male', 5);
+--echo # Correct usage of NTILE with a fix argument NTILE(6). 
+select c1, c2, c3, ntile(6) over (partition by c2 order by pk) from t1;
+--echo # Correct usage - constant NTILE (argument) in each partition.
```
Comment (an3l) [inline mysql-test/t/win_ntile.test:207]: Whitespace

File: sql/item_windowfunc.h [hunk 7]
```diff
@@ -618,7 +618,8 @@ class Item_sum_ntile : public Item_sum_window_with_row_count
  public:
   Item_sum_ntile(THD* thd, Item* num_quantiles_expr) :
     Item_sum_window_with_row_count(thd, num_quantiles_expr),
-    current_row_count_(0) {};
+    current_row_count_(0),
```
Comment (an3l) [inline sql/item_windowfunc.h:621]: Whitespace

File: sql/item_windowfunc.h [hunk 8]
```diff
@@ -683,6 +686,7 @@ class Item_sum_ntile : public Item_sum_window_with_row_count
  private:
   longlong get_num_quantiles() { return args[0]->val_int(); }
   ulong current_row_count_;
+  longlong n_old_val_;
```
Comment (an3l) [inline sql/item_windowfunc.h:689]: Why `longlong`, can it ever be negative?
I would prefer here `ulonglong`.

Comment (DanAtSPB) [reply sql/item_windowfunc.h:689]: Then we not allowed to compare num_quantiles != n_old_val_

Comment (an3l) [reply sql/item_windowfunc.h:689]: Then solve it with casting.

File: sql/item_windowfunc.h (hunk 1 above)
Comment (DanAtSPB) [reply sql/item_windowfunc.h]: n_old_val > 0 implemented

File: sql/item_windowfunc.h (hunk 8 above)
Comment (DanAtSPB) [reply sql/item_windowfunc.h:689]: I've used reinterpret_cast <longlong> (n_old_val_) 
Does it look fine?

Comment (DanAtSPB) [reply sql/item_windowfunc.h:689]: May I use 
auto n_old_val_   
with c++11 set(CMAKE_CXX_STANDARD 11)

Comment (DanAtSPB) [reply sql/item_windowfunc.h]: I've used static_cast.  Does it look fine?

**Comment by montywi** [comment]
Hi!

On Tue, 23 Jun 2020, 14:20 DanAtSPB, <notifications@github.com> wrote:

> *@DanAtSPB* commented on this pull request.
> ------------------------------
>
> In sql/item_windowfunc.h
> <https://github.com/MariaDB/server/pull/1596#discussion_r444148853>:
>
> > @@ -683,6 +686,7 @@ class Item_sum_ntile : public Item_sum_window_with_row_count
>   private:
>    longlong get_num_quantiles() { return args[0]->val_int(); }
>    ulong current_row_count_;
> +  longlong n_old_val_;
>
> May I use
> auto n_old_val_
> with c++11 set(CMAKE_CXX_STANDARD 11).
>
Please don't use "auto". It makes it much harder to review and understand
the code as you don't know the type of the variable just by looking at the
code!
"auto" can also cause invisible conversion problems later in the code when
the variable is used as parameters to other functions, like printf.

Regards,
Monty

**Comment by DanAtSPB** [comment]
> "auto" can also cause invisible conversion problems later in the code when the variable is used as parameters to other functions, like printf
I've used "auto" before, but this time ulonglong used
and because num_quantiles is longlong it looks a little bit longer at the last version:
n_old_val_= static_cast<ulonglong>(num_quantiles);
https://github.com/MariaDB/server/pull/1596/files

**Pull request merged**